// cdk-nag/lib/rules/iam/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IAMUserNoPolicies = exports.IAMUserGroupMembership = exports.IAMPolicyNoStatementsWithFullAccess = exports.IAMPolicyNoStatementsWithAdminAccess = exports.IAMNoWildcardPermissions = exports.IAMNoManagedPolicies = exports.IAMNoInlinePolicy = exports.IAMGroupHasUsers = void 0;
/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/
var IAMGroupHasUsers_1 = require("./IAMGroupHasUsers");
Object.defineProperty(exports, "IAMGroupHasUsers", { enumerable: true, get: function () { return IAMGroupHasUsers_1.default; } });
var IAMNoInlinePolicy_1 = require("./IAMNoInlinePolicy");
Object.defineProperty(exports, "IAMNoInlinePolicy", { enumerable: true, get: function () { return IAMNoInlinePolicy_1.default; } });
var IAMNoManagedPolicies_1 = require("./IAMNoManagedPolicies");
Object.defineProperty(exports, "IAMNoManagedPolicies", { enumerable: true, get: function () { return IAMNoManagedPolicies_1.default; } });
var IAMNoWildcardPermissions_1 = require("./IAMNoWildcardPermissions");
Object.defineProperty(exports, "IAMNoWildcardPermissions", { enumerable: true, get: function () { return IAMNoWildcardPermissions_1.default; } });
var IAMPolicyNoStatementsWithAdminAccess_1 = require("./IAMPolicyNoStatementsWithAdminAccess");
Object.defineProperty(exports, "IAMPolicyNoStatementsWithAdminAccess", { enumerable: true, get: function () { return IAMPolicyNoStatementsWithAdminAccess_1.default; } });
var IAMPolicyNoStatementsWithFullAccess_1 = require("./IAMPolicyNoStatementsWithFullAccess");
Object.defineProperty(exports, "IAMPolicyNoStatementsWithFullAccess", { enumerable: true, get: function () { return IAMPolicyNoStatementsWithFullAccess_1.default; } });
var IAMUserGroupMembership_1 = require("./IAMUserGroupMembership");
Object.defineProperty(exports, "IAMUserGroupMembership", { enumerable: true, get: function () { return IAMUserGroupMembership_1.default; } });
var IAMUserNoPolicies_1 = require("./IAMUserNoPolicies");
Object.defineProperty(exports, "IAMUserNoPolicies", { enumerable: true, get: function () { return IAMUserNoPolicies_1.default; } });
